<a href="https://colab.research.google.com/github/PriyathamVarma/PriyathamVarma/blob/main/IDDFS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [105]:
# The solution to the 8-tile problem involves 4 steps.
# Step 1 : Classes and its methods for start and goal states tracking while the process of search.
# Step 2 : Classes and its methods related to solving the puzzle.
# Step 3 : Classes and methods for outputting the result
# Step 4 : Classes and methods for the main strategy which is Iterative deepening depth first search
# In the last all the classes will be utilised to get the required output.

# Imports
import numpy as np
import time
import itertools


class Vertex:
# Initial method declaration
    def __init__(self, start_state, goal_state=None):
        self.start_state = start_state
        self.goal_state = goal_state
# Method for sequence
    @property
    def seq(self): # to keep track of the sequence used to get to the goal
        node, seq = self, []
        while node:
            seq.append(node)
            node = node.goal_state
        yield from reversed(seq)
# Method for state
    @property
    def state(self):
        return str(self.start_state.board) # hashable so it can be compared in sets
# Method for solved puzzle
    @property
    def isSolved(self):
        return self.start_state.isSolved
# Method for getting moves
    @property
    def getMoves(self):
        return self.start_state.getMoves
# class declaration for puzzle solution
class Puzzle:
# initial method declaration for puzzle class
    def __init__(self, startBoard):
        self.board = startBoard
# method for getting moves
    @property
    def getMoves(self):

        possibleNewBoards = []

        zeroPos = self.board.index(0) # find the zero tile to determine possible moves
# shuffling the puzzle
        if zeroPos == 0:
            possibleNewBoards.append(self.move(0,1))
            possibleNewBoards.append(self.move(0,3))
        elif zeroPos == 1:
            possibleNewBoards.append(self.move(1,0))
            possibleNewBoards.append(self.move(1,2))
            possibleNewBoards.append(self.move(1,4))
        elif zeroPos == 2:
            possibleNewBoards.append(self.move(2,1))
            possibleNewBoards.append(self.move(2,5))
        elif zeroPos == 3:
            possibleNewBoards.append(self.move(3,0))
            possibleNewBoards.append(self.move(3,4))
            possibleNewBoards.append(self.move(3,6))
        elif zeroPos == 4:
            possibleNewBoards.append(self.move(4,1))
            possibleNewBoards.append(self.move(4,3))
            possibleNewBoards.append(self.move(4,5))
            possibleNewBoards.append(self.move(4,7))
        elif zeroPos == 5:
            possibleNewBoards.append(self.move(5,2))
            possibleNewBoards.append(self.move(5,4))
            possibleNewBoards.append(self.move(5,8))
        elif zeroPos == 6:
            possibleNewBoards.append(self.move(6,3))
            possibleNewBoards.append(self.move(6,7))
        elif zeroPos == 7:
            possibleNewBoards.append(self.move(7,4))
            possibleNewBoards.append(self.move(7,6))
            possibleNewBoards.append(self.move(7,8))
        else:
            possibleNewBoards.append(self.move(8,5))
            possibleNewBoards.append(self.move(8,7))

        return possibleNewBoards # returns Puzzle objects (maximum of 4 at a time)
# method for moving places/positions
    def move(self, current, to):

        changeBoard = self.board[:] # create a copy
        changeBoard[to], changeBoard[current] = changeBoard[current], changeBoard[to] # switch the tiles at the passed positions
        return Puzzle(changeBoard) # return a new Puzzle object
# method for printing puzzle
    def printPuzzle(self): # prints board in 8 puzzle style

        copyBoard = self.board[:]
        """for i in range(9):
            if i == 2 or i == 5:
                print((str)(copyBoard[i]))
            else:
                print((str)(copyBoard[i])+" ", end="")
                
        print('\n')"""
# method for getting solved
    @property
    def isSolved(self):
        return self.board == [0,1,2,3,4,5,6,7,8] # goal board
# declaring class for solving puzzle
class Solver:
# initial method declaration
    def __init__(self, Puzzle):
        self.start_state = Puzzle
# method for the data structure
    def IDDFS(self):

        def DLS(currentNode, depthSearch):
            if depthSearch == 0:
                return None
            if currentNode.isSolved:
                return currentNode
            elif depthSearch > 0:
                for board in currentNode.getMoves:
                    nextNode = Vertex(board, currentNode)
                    if nextNode.state not in visited:
                        visited.add(nextNode.state)
                        goalNode = DLS(nextNode, depthSearch - 1)
                        if goalNode != None: # I thought this should be redundant but it never finds a soln if I take it out
                            if goalNode.isSolved: # same as above ^
                                return goalNode

        for depthSearch in itertools.count():
            visited = set()
            startNode = Node(self.start_state)
            #print(startNode.isSolved)
            goalNode = DLS(startNode, depthSearch)
            if goalNode != None:
                if goalNode.isSolved:
                    return goalNode.seq

Goal_state = [0, 2, [[3, 2, 0], [6, 1, 8], [4, 7, 5]]]
print("GOAL STATE:",Goal_state)
print("***********************")

main_array = [[0, 0, [[0, 7, 1], [4, 3, 2], [8, 6, 5]]],
[0, 2, [[5, 6, 0], [1, 3, 8], [4, 7, 2]]],
[2, 0, [[3, 5, 6], [1, 2, 7], [0, 8, 4]]],
[1, 1, [[7, 3, 5], [4, 0, 2], [8, 1, 6]]],
[2, 0, [[6, 4, 8], [7, 1, 3], [0, 2, 5]]]]

for k in main_array:
  required_array = k
  startingBoard = np.array(required_array[2])

  startingBoard2 = startingBoard.tolist()


  array_list = []
  for i in range(0,len(startingBoard2)):
    for j in range(0,len(startingBoard2)):
      array_list.append(startingBoard2[i][j])


  myPuzzle = Puzzle(array_list)
  mySolver = Solver(myPuzzle)
  start = time.time()
  goalSeq = mySolver.IDDFS()
  end = time.time()

  counter = -1 # starting state doesn't count as a move
  for node in goalSeq:
      counter = counter + 1
      node.start_state.printPuzzle()
      


  data = np.array(startingBoard)
  reshaped = data.reshape(3,3)

  result = np.where(reshaped == 0)

  print("RESULTS FOR:")
  print("[",result[0],",",result[1],",")
  print(reshaped)
  print("NUMBER OF MOVES TO SOLVE: " + str(counter))
  totalTime = end - start
  print("THE COMPUTING TIME THE SEARCH TOOK: %.2f seconds" % (totalTime))
  print("---------------------------------------------------------------")

GOAL STATE: [0, 2, [[3, 2, 0], [6, 1, 8], [4, 7, 5]]]
***********************
RESULTS FOR:
[ [0] , [0] ,
[[0 7 1]
 [4 3 2]
 [8 6 5]]
NUMBER OF MOVES TO SOLVE: 22
THE COMPUTING TIME THE SEARCH TOOK: 0.20 seconds
---------------------------------------------------------------
RESULTS FOR:
[ [0] , [2] ,
[[5 6 0]
 [1 3 8]
 [4 7 2]]
NUMBER OF MOVES TO SOLVE: 34
THE COMPUTING TIME THE SEARCH TOOK: 3.12 seconds
---------------------------------------------------------------
RESULTS FOR:
[ [2] , [0] ,
[[3 5 6]
 [1 2 7]
 [0 8 4]]
NUMBER OF MOVES TO SOLVE: 36
THE COMPUTING TIME THE SEARCH TOOK: 3.45 seconds
---------------------------------------------------------------
RESULTS FOR:
[ [1] , [1] ,
[[7 3 5]
 [4 0 2]
 [8 1 6]]
NUMBER OF MOVES TO SOLVE: 28
THE COMPUTING TIME THE SEARCH TOOK: 0.94 seconds
---------------------------------------------------------------
RESULTS FOR:
[ [2] , [0] ,
[[6 4 8]
 [7 1 3]
 [0 2 5]]
NUMBER OF MOVES TO SOLVE: 30
THE COMPUTING TIME THE SEARCH TOOK: 1.50 seconds
-

GOAL STATE: [0, 2, [[3, 2, 0], [6, 1, 8], [4, 7, 5]]]
***********************
RESULTS FOR:
[ [0] , [0] ,
[[0 7 1]
 [4 3 2]
 [8 6 5]]
NUMBER OF MOVES TO SOLVE: 22
THE COMPUTING TIME THE SEARCH TOOK: 0.20 seconds
---------------------------------------------------------------
RESULTS FOR:
[ [0] , [2] ,
[[5 6 0]
 [1 3 8]
 [4 7 2]]
NUMBER OF MOVES TO SOLVE: 34
THE COMPUTING TIME THE SEARCH TOOK: 3.12 seconds
---------------------------------------------------------------
RESULTS FOR:
[ [2] , [0] ,
[[3 5 6]
 [1 2 7]
 [0 8 4]]
NUMBER OF MOVES TO SOLVE: 36
THE COMPUTING TIME THE SEARCH TOOK: 3.45 seconds
---------------------------------------------------------------
RESULTS FOR:
[ [1] , [1] ,
[[7 3 5]
 [4 0 2]
 [8 1 6]]
NUMBER OF MOVES TO SOLVE: 28
THE COMPUTING TIME THE SEARCH TOOK: 0.94 seconds
---------------------------------------------------------------
RESULTS FOR:
[ [2] , [0] ,
[[6 4 8]
 [7 1 3]
 [0 2 5]]
NUMBER OF MOVES TO SOLVE: 30
THE COMPUTING TIME THE SEARCH TOOK: 1.50 seconds
---------------------------------------------------------------